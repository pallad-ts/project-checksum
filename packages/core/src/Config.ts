import {Algorithm} from "./Algorithm";

export interface RawConfig {
	readonly algorithm: Algorithm;
	readonly paths?: string[]
	readonly external?: RawConfig.External[];
}

export namespace RawConfig {
	export type External = string | { loader: string, options: unknown };

	export interface Enchanted extends RawConfig {
		readonly directory: string;
	}
}
