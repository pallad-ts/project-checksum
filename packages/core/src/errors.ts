import {Domain, generators} from "alpha-errors";
import {ChecksumError} from "./ChecksumError";

export const ERRORS = Domain.create({
	errorClass: ChecksumError,
	codeGenerator: generators.formatCode('E_PROJECT_CHECKSUM_%d')
})
	.createErrors(create => {
		return {
			NO_CONFIG_FOUND: create('No @pallad/project-checksum configuration found in directory: %s'),
			NO_CONFIG_FILE_FOUND: create('No config file "%s" for @pallad/project-checksum found in directory: %s'),
			NO_LOADER_FOUND: create('No loader "%s" found'),
			NO_LOADER_EXPORTED: create('No loader exported in "%s" package'),
			NO_CONTENT_TO_HASH_COMPUTE_FOUND: create("Project does not have any content to compute checksum from"),
			CIRCULAR_DEPENDENCY: create('Circular dependency detected. Projects path: %s')
		}
	});
