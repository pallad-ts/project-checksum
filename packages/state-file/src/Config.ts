import {ProjectConfig} from "./types";

export interface Config {
	readonly stateFile: string;
	readonly projects: Record<string, ProjectConfig>;
}
