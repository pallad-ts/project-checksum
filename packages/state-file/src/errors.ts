import {Domain, generators} from "alpha-errors";
import {ChecksumError} from "@pallad/project-checksum-core";

export const ERRORS = Domain.create({
	errorClass: ChecksumError,
	codeGenerator: generators.formatCode('E_PROJECT_CHECKSUM_STATE_%d')
})
	.createErrors(create => {
		return {
			NO_PROJECT_FOUND: create('No checksum project "%s" found'),
			STATE_FILE_NOT_FOUND: create('No state file "%s" found'),
			NO_PROJECT_CHECKSUM_IN_STATE: create('No checksum in state for project "%s"'),
			NO_CONFIG_FOUND: create('No configuration file found in directory "%s"'),
			CONFIG_FILE_NOT_FOUND: create('Configuration file "%s" not found')
		}
	});
