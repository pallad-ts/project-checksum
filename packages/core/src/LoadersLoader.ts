import {ERRORS} from "./errors";
import {Loader} from "./Loader";
import {Either, left, right} from "@sweet-monads/either";

// eslint-disable-next-line @typescript-eslint/naming-convention
function fromTry<L, R>(fn: () => R): Either<L, R> {
	try {
		return right(fn());
	} catch (e) {
		return left(e as L);
	}
}

export class LoadersLoader {
	private loaders = new Map<string, Loader<unknown>>();

	getLoader(name: string): Loader<unknown> {
		if (this.loaders.has(name)) {
			return this.loaders.get(name)!
		}

		const packageNames = [`@pallad/project-checksum-loader-${name}`, name];
		for (const packageName of packageNames) {
			const module = this.tryLoadModule(packageName);
			if (module) {
				const loader = this.createLoaderFromModule(module, packageName);
				this.loaders.set(name, loader);
				return loader;
			}
		}
		throw ERRORS.NO_LOADER_FOUND.format(name);
	}

	private createLoaderFromModule(module: unknown, packageName: string) {
		// eslint-disable-next-line no-null/no-null
		if (typeof module === 'object' && module !== null && 'Loader' in module) {
			const loader = (module as any).Loader;
			if (typeof loader === 'function') {
				return new loader();
			}
		}
		throw ERRORS.NO_LOADER_EXPORTED.format(packageName);
	}

	private tryLoadModule(name: string) {
		return fromTry(() => require(name))
			.mapLeft(() => undefined).value;
	}
}
