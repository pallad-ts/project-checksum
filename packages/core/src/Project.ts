import {Dependency} from "./Dependency";
import {createHashForAlgorithm} from "./createHashForAlgorithm";
import {Algorithm} from "./Algorithm";
import {ERRORS} from "./errors";

export class Project {
	constructor(readonly config: Project.Config) {
	}

	async computeChecksum(): Promise<string> {
		const hash = createHashForAlgorithm(this.config.algorithm);
		let found = false;
		for (const dependency of this.config.dependencies) {
			const dependencyContent = await dependency.getContent();
			const isRegularContent = Buffer.isBuffer(dependencyContent) || Array.isArray(dependencyContent);
			const finalContent = isRegularContent ? [dependencyContent] : dependencyContent;
			for await (const content of finalContent) {
				if (!content) {
					continue;
				}
				found = true;
				if (Buffer.isBuffer(content)) {
					hash.update(content);
				} else {
					const [string, encoding] = content;
					hash.update(string, encoding);
				}
			}
		}

		if (!found) {
			throw ERRORS.NO_CONTENT_TO_HASH_COMPUTE_FOUND();
		}
		return hash.digest('hex');
	}
}

export namespace Project {
	export interface Config {
		algorithm: Algorithm;
		dependencies: Dependency[],
		rootDirectory: string
	}
}
