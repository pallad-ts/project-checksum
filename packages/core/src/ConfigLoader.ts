import {SchemaValidation, Validation, ValidatorError, ViolationsList} from "alpha-validator";
import {ERRORS} from "./errors";
import {Config} from "./Config";
import {cosmiconfig} from "cosmiconfig";
import * as path from "path";
import {byJsonSchema} from "alpha-validator-bridge-jsonschema";
import {LoadersLoader} from "./LoadersLoader";
import {Project} from "./Project";
import {Dependency, FilesPatternsDependency} from "./Dependency";

const validation = SchemaValidation.toValidationFunction(
	'none', byJsonSchema<Config>(require('../config-schema.json'))
);

export class ConfigLoader {
	private explorer = cosmiconfig('@pallad/project-checksum', {
		searchPlaces:
			[
				'@pallad/project-checksum',
				'pallad-project-checksum',
				'pallad-project-checksum.config',
				'project-checksum'
			]
				.reduce((result, name) => {
					return result.concat([
						`${name}.json`,
						`${name}.yaml`,
						`${name}.yml`,
						`${name}.js`,
					])
				}, [] as string[]),
	});

	constructor(private loadersLoader: LoadersLoader) {
	}

	async validate(data: any): Promise<Validation<ViolationsList, Config>> {
		return (await validation(data))
			.chain(config => {
				const hasPaths = Array.isArray(config.paths) && config.paths.length > 0;
				const external = Array.isArray(config.external) && config.external.length > 0;
				if (!hasPaths && !external) {
					return Validation.Fail(
						ViolationsList.create()
							.addViolation('Project configuration needs at least one path or external defined'),
					);
				}
				return Validation.Success(config);
			});
	}

	async loadProjectFromFile(file: ConfigLoader.ConfigFile, context?: ConfigLoader.Context): Promise<Project> {
		const fullFilePath = path.join(file.rootDir, file.filePath);

		if (context && context.loadedFiles.includes(fullFilePath)) {
			throw ERRORS.CIRCULAR_DEPENDENCY.format(
				context.loadedFiles.concat([fullFilePath]).join(' > ')
			);
		}

		const result = await this.explorer.load(fullFilePath);
		if (!result) {
			throw ERRORS.NO_CONFIG_FILE_FOUND.format(file.filePath, file.rootDir);
		}

		const config = await this.validateConfigOrFail(result.config);
		const finalContext: ConfigLoader.Context = context ?? {loadedFiles: [fullFilePath]};
		return this.createProjectFromConfig(config, file, finalContext);
	}

	private async createProjectFromConfig(config: Config, file: ConfigLoader.ConfigFile, context: ConfigLoader.Context) {
		const dependencies = [] as Dependency[];
		if (config.paths && config.paths.length > 0) {
			dependencies.push(new FilesPatternsDependency(config.paths, file.rootDir));
		}

		if (config.external) {
			for (const external of config.external) {
				const {loader: loaderName, options: loaderOptions} = Config.External.toLoaderConfig(external);
				const loader = this.loadersLoader.getLoader(loaderName);

				const loadingResult = await loader.load({
					configLoader: this,
					loadedFiles: context.loadedFiles,
					directory: file.rootDir
				}, loaderOptions);

				if (loadingResult) {
					dependencies.push(
						...(Array.isArray(loadingResult) ? loadingResult : [loadingResult])
					);
				}
			}
		}

		return new Project({
			algorithm: config.algorithm,
			dependencies,
			rootDirectory: file.rootDir
		});
	}

	private async validateConfigOrFail(config: unknown) {
		const validationResult = await this.validate(config);
		if (validationResult.isFail()) {
			throw new ValidatorError(validationResult.fail(), 'Invalid config');
		}
		return validationResult.success();
	}

	async loadProjectFromDirectory(directory: string): Promise<Project> {
		const result = await this.explorer.search(directory);
		if (!result) {
			throw ERRORS.NO_CONFIG_FOUND.format(directory);
		}

		return this.loadProjectFromFile({
			filePath: path.relative(directory, result.filepath),
			rootDir: directory
		});
	}
}

export namespace ConfigLoader {
	export interface Context {
		loadedFiles: string[];

	}

	export interface ConfigFile {
		filePath: string;
		rootDir: string;
	}
}
