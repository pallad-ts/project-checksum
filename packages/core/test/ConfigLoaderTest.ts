import {Config} from "@src/Config";
import * as path from "path";
import {ERRORS} from "@src/errors";
import {ConfigLoader} from "@src/ConfigLoader";
import {LoadersLoader} from "@src/LoadersLoader";
import * as sinon from 'sinon';
import {Project} from "@src/Project";
import {left, right} from "@sweet-monads/either";

describe('ConfigLoader', () => {
	const DATA: Config = {
		algorithm: 'md5',
		paths: ['some-file-path']
	};

	let configLoader: ConfigLoader;
	let loadersLoader: sinon.SinonStubbedInstance<LoadersLoader>;

	beforeEach(() => {
		loadersLoader = sinon.createStubInstance(LoadersLoader);
		configLoader = new ConfigLoader(loadersLoader as unknown as LoadersLoader);
	});

	describe('validation', () => {
		it('valid', () => {

		});

		describe('invalid', () => {
			it('algorithm', async () => {
				// eslint-disable-next-line @typescript-eslint/await-thenable
				const result = await configLoader.validate({
					...DATA,
					algorithm: 'test'
				});
				expect(result.value)
					.toMatchSnapshot()
			});

			it('algorithm type', async () => {
				// eslint-disable-next-line @typescript-eslint/await-thenable
				const result = await configLoader.validate({
					...DATA,
					algorithm: 4
				});
				expect(result.value)
					.toMatchSnapshot();
			});

			it('missing paths and external', async () => {
				// eslint-disable-next-line @typescript-eslint/await-thenable
				const result = await configLoader.validate({
					...DATA,
					paths: [],
					external: []
				});
				expect(result.value)
					.toMatchSnapshot();
			})
		})
	});

	describe('loading config', () => {
		it('from yml file', async () => {
			const directory = path.resolve(__dirname, './fixtures/projects/test-yml');
			const project = await configLoader.loadProjectFromDirectory(directory);

			expect(project)
				.toBeInstanceOf(Project);

			expect(project.config.dependencies)
				.toHaveLength(1);
		});

		it('fails if no config was found', async () => {
			const directory = path.resolve(__dirname, './fixtures/projects/no-config');
			const config = await configLoader.loadProjectFromDirectory(directory).then(right, left);
			expect(ERRORS.NO_CONFIG_FOUND.is(config.value))
				.toBeTruthy();
		});

		describe('fails', () => {
			it('when circular dependency detected', async () => {
				const result = await configLoader.loadProjectFromFile({
					rootDir: 'a',
					filePath: 'file.yml'
				}, {
					loadedFiles: [
						'a/file.yml',
						'b.file.yml'
					]
				}).then(right, left);

				expect(ERRORS.CIRCULAR_DEPENDENCY.is(result.value))
					.toBeTruthy();

				expect(result.value.message)
					.toMatchInlineSnapshot(`"Circular dependency detected. Projects path: a/file.yml > b.file.yml > a/file.yml"`);
			})
		})
	});
});
