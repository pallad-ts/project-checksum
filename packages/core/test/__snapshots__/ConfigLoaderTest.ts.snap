// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigLoader validation invalid algorithm 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "must be equal to one of the allowed values",
      "path": [
        "algorithm",
      ],
    },
  ],
}
`;

exports[`ConfigLoader validation invalid algorithm type 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "must be string",
      "path": [
        "algorithm",
      ],
    },
  ],
}
`;

exports[`ConfigLoader validation invalid missing paths and external 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "must contain at least 1 valid item(s)",
      "path": [
        "paths",
      ],
    },
  ],
}
`;
